### ISSUE TITLE
presetDate not work with range disabled date

### ISSUE THREAD

**Issue by impetoyt** [WhatSock/apex#51]
I upload the code here:
[https://easyupload.io/ogl4ic](https://easyupload.io/ogl4ic)

1) open the calendar and choose a date from the range (take note that is the year 2023)
2) open the browser console and write: `$A("departure_picker").presetDate($A("departure_picker"), new Date("2020-12-09"), new Date("2020-11-11"), new Date("2020-12-12"));`
3) open again the calendar. Everything is set fine except for the years selection dropdown that is completly empty and doesn't show the current year which is 2020
4) try to write again in the console this `$A("departure_picker").presetDate($A("departure_picker"), new Date("2023-12-09"), new Date("2023-11-11"), new Date("2023-12-12"));`
basically you are again choose the year 2023. Open the calendar and you will see that the years selection dropdown is working again and it also show the year which is 2023.

**Comment by impetoyt** [comment]
I found the problem, the difference between my code and yours is that I use these other two properties in the calendar configuration:
```
		yearSelectMin: (new Date("2023-04-05")).getFullYear()
		yearSelectMax: (new Date("2023-05-25")).getFullYear()
```

So when I use the presetDate method which change minDate and maxDate it appear the issue just for the years selection since for the month dosn't exist that two properties. So in another word:
`$A("departure_picker").presetDate($A("departure_picker"), new Date("2020-12-09"), new Date("2020-11-11"), new Date("2020-12-12"));`

The presetDate change correctly the minDate and maxDate but since the yearSelectMin or yearSelectMax is still set on the year 2023 the dropdown selection is empty

**Comment by accdc** [comment]
Hi, I've been looking into this and the issue with your code is not actually the javascript, it is whatever changes you have made to your datepicker.css file.

When I copy datepicker.css from the 4X repo and paste that into your example code folder, instantly the functionality starts working as expected.

I noticed this because when you run the datepicker from your sample code, it is not possible to tab between the various controls via Tab or Shift+Tab, and the hidden select elements are not being hidden properly which is breaking the functionality. This isn't a JS issue but a CSS issue within your sample code.

**Comment by impetoyt** [comment]
In my Datapicker.css I put display none and display in-line in order to show always the selection field for the years and months in order to allow the users to avoid two clicks to open the relative dropdowns. You said that it was a compromise that the user should click two times but from the UX point of view isn't good. Before getting again into this we can easly reproduce the issue which is js problem not css in your test code:
[https://whatsock.com/Templates/Datepickers/With%20Absolute%20Disabled%20Date%20Ranges/index.htm](https://whatsock.com/Templates/Datepickers/With%20Absolute%20Disabled%20Date%20Ranges/index.htm)

1) choose a date from the calendar;
2) in the browser console: `$A("UniqueCalendarId").presetDate($A("UniqueCalendarId"), new Date("1800-12-09"), new Date("1799-12-09"), new Date("1802-12-09"));`
3) open the calendar again: you can't click on the previous year `<td class="nav prev btn year disabled"...` since is disabled, you can click on the years selection but the dropdown is empty. I notice that this happens because in your test page you have the follow configuration for the calendar:
`    yearSelectMin: 1900,
    yearSelectMax: 2030,`

So when you change the minDate and maxDate with presetDate and you put a years range that it isn't inside the range of yearSelectMin and yearSelectMax the issue appear.

**Comment by accdc** [comment]
I understand that if you have a datepicker that is preconfigured for a specific year date range, then you change it so that the new range is outside of that range then the select will not show any options.

This will happen if you hack a datepicker, but I don't understand why this would ever happen in practice for anyone actually using a datepicker to select a date. Why would you want to do this on a webpage?

To fix this within your datepicker, simply set the following:

```
yearSelectMin: 1,
yearSelectMax: 9999,
```

This will work with any reasonable min max date range.

Regarding your CSS, if you choose to force the select to be visible and the button to be hidden, it will break accessibility for all non-sighted screen reader users.

**Comment by impetoyt** [comment]
Not an hack. I encounter this problem when I used three calendar. When you choose a date from the first two calendar than the third calendar is setted with minDate and yearSelectMin of the first calendar date and maxDate and yearSelectMax of the second calendar date.
So this is the exact example where when you change again the date of the first two calendar the third one could encounter the issue. Again as I said and as you said the solution is to put on the third calendar something like indipendent from the first two calendar:
`yearSelectMin: 1,
yearSelectMax: 9999,`

Regarding the css, it wasn't my intention to break the accessibility which is the most important things to me (I'll restore my css file). I tried to modify the css since many users was not comfortable to click two times in order to choose the month or the years.

**Comment by accdc** [comment]
Hi, as a quick note, I believe I found a way of rendering the select so that it doesn't cause any issues for accessibility. I have most of it done, but will have it ready next week. I'll reach out when it's ready.

**Comment by accdc** [comment]
I've updated the Datepicker module to version 5.1 which now includes the ability to render just the year/month select element instead of a button. 
https://whatsock.com/Templates/Datepickers/Basic/index.htm

This requires setting the following property in your config file:

```
// Force the month/year select dropdown to render instead of a button.
forceSelect: true,
```

**Comment by accdc** [comment]
I also just pushed an update to Datepicker 5.2 that will auto-adjust the yearSelectMin and yearSelectMax values whenever it detects that the minDate or maxDate properties have been changed or are out of sync.
